gazelle_haskell_modules/rule_generation.go

```diff
@@ -278,36 +278,43 @@ func concatRuleInfos(xs [][]*RuleInfo) []*RuleInfo {
 // 0. The files listed in the srcs attribute of a rule
 // 1. All the files (recursively) in directories specified in a "gazelle_haskell_modules:srcs:"
 func getSrcs(pkgRoot string, r *rule.Rule) ([]string, error) {
-	var srcs []string
-
 	sourcesFromRule, err := srcsFromRule(pkgRoot, r)
 	if err != nil {
 		return nil, fmt.Errorf("getSrcs: %w", err)
 	}
-	srcs = append(srcs, sourcesFromRule...)
 
 	autodiscoveredSrcs, err := autodiscoverSrcs(pkgRoot, r)
 	if err != nil {
 		return nil, fmt.Errorf("getSrcs: %w", err)
 	}
-	srcs = append(srcs, autodiscoveredSrcs...)
+
+	// reuse the set we got form srcsFromRule to deduplicate autodiscovered and explicit files
+	for _, v := range autodiscoveredSrcs {
+		sourcesFromRule[v] = true
+	}
+
+	var srcs []string
+	for file := range sourcesFromRule {
+		srcs = append(srcs, file)
+	}
```

Comment:
Do you know if the order in which srcs are returned is deterministic? This could affect the order in which rules are generated, maybe.

Reply:
Also, I've read about pre-allocating an array with the final size, rather than resizing it with `n` appends. cc @YorikSar

Reply:
> Do you know if the order in which srcs are returned is deterministic?

It is not. But there are other sources of non-determinism in `gazelle_haskell_modules` too. I think it is fair to defer on this.